File: packages/php/src/Resources/Contacts.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `contacts` resource: phone contacts with free-form `attributes`. Fields: `e164`, `name`, `attributes`.
 * Accessed as `$client->contacts`.
 */
final class Contacts
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Query: `limit` (1..200), `cursor`.
     *
     * @param array<string, mixed> $params
     */
    public function list(array $params = []): Page
    {
        $query = Models::map($params, Models::PAGE_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/contacts', $query));
    }

    /**
     * @param array<string, mixed> $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function create(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::CONTACT, ['e164']);

        return Models::object($this->transport->request('POST', '/v1/contacts', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/contacts/' . Models::pathParam($id)));
    }

    /**
     * Partial update (PATCH): only the given fields change.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function update(string $id, array $params): array
    {
        $body = Models::map($params, Models::CONTACT);

        return Models::object($this->transport->request('PATCH', '/v1/contacts/' . Models::pathParam($id), [], $body));
    }

    public function delete(string $id): void
    {
        $this->transport->request('DELETE', '/v1/contacts/' . Models::pathParam($id));
    }
}

File: packages/java/src/main/java/io/opensms/Contacts.java
package io.opensms;

import io.opensms.models.Contact;
import io.opensms.models.Page;

/** Manage contacts. Accessed as {@code client.contacts()}. */
public final class Contacts {

    private final ApiTransport t;

    Contacts(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of contacts. */
    public Page<Contact> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<Contact> list(ListParams params) {
        return t.get("/v1/contacts", params.toQuery(), ApiTransport.pageOf(Contact.class));
    }

    /**
     * @param params {@code e164} required.
     * @return the created contact.
     */
    public Contact create(ContactParams params) {
        return create(params, null);
    }

    /**
     * @param params  {@code e164} required.
     * @param options per-call options.
     * @return the created contact.
     */
    public Contact create(ContactParams params, RequestOptions options) {
        Wire.require(params.e164(), "e164");
        return t.json("POST", "/v1/contacts", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(Contact.class));
    }

    /**
     * @param id contact id.
     * @return the contact.
     */
    public Contact get(String id) {
        return t.get("/v1/contacts/" + Wire.id(id, "id"), null, ApiTransport.type(Contact.class));
    }

    /**
     * Partial update; fields you do not set are kept.
     *
     * @param id     contact id.
     * @param params fields to change.
     * @return the updated contact.
     */
    public Contact update(String id, ContactParams params) {
        return t.json("PATCH", "/v1/contacts/" + Wire.id(id, "id"), params.toWire(), null,
                ApiTransport.type(Contact.class));
    }

    /** @param id contact id. */
    public void delete(String id) {
        t.json("DELETE", "/v1/contacts/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }
}
